nipap/nipap/authlib.py

```diff
@@ -271,6 +287,107 @@ def authorize(self):
+                Authoritative source of the query.
+            * `auth_options` [dict]
+                A dict which, if authenticated as a trusted user, can override
+                `username` and `authoritative_source`.
+        """
+
+        if auth_options is None:
+            auth_options = {}
+
+        BaseAuth.__init__(self, None, None, authoritative_source,
+                          name, auth_options)
+
+        self._jwt_token = jwt_token
+
+        base_auth_backend = 'auth.backends.' + self.auth_backend
+        if self._cfg.has_option(base_auth_backend, 'ro_group'):
+            self._jwt_ro_group = self._cfg.get(base_auth_backend, 'ro_group')
+        if self._cfg.has_option(base_auth_backend, 'rw_group'):
+            self._jwt_rw_group = self._cfg.get(base_auth_backend, 'rw_group')
+
+    def authenticate(self):
+        """ Verify authentication.
+
+            Returns True/False dependant on whether the authentication
+            succeeded or not.
+        """
+
+        # if authentication has been performed, return last result
+        if self._authenticated is not None:
+            return self._authenticated
+
+        try:
+            self._token = self._cfg.get('auth.backends.' +
+                                        self.auth_backend, 'jwk_url')
+            # Fetch JWKs (done when initializing JwtAuth-class),
+            # keep the keys in the class instance
+            jwk_request_response = requests.get(self._token)
+            jwks = jwk_request_response.json()
+            jwk_keys = {}
+            for jwk in jwks['keys']:
+                kid = jwk['kid']
+                jwk_keys[kid] = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk))
+
+            # # Upon auth with a JWT-token
+            # # Retrieve key for token
+            jwt_headers = jwt.get_unverified_header(self._jwt_token)
+            jwt_jwk_key = jwk_keys[jwt_headers['kid']]
+
+            # # Decode and verify token
+            payload = jwt.decode(
+                self._jwt_token,
+                key=jwt_jwk_key,
+                algorithms=[jwt_headers['alg']],
+                options={"verify_aud": False})
+
+            # # Setting read and write rights
+            if payload.get('groups'):
+                if self._jwt_ro_group in payload.get('groups'):
+                    self.readonly = True
+                    self._authenticated = True
+                elif self._jwt_rw_group in payload.get('groups'):
+                    self.readonly = False
+                    self._authenticated = True
+        except:
+            # Auth failed
+            self._logger.debug('could not authenticate')
+            self._authenticated = False
+            return self._authenticated
+
+        # auth succeeded
+        if self._authenticated:
+            self.username = payload.get('sub')
+            self.authenticated_as = payload.get('sub')
+            self.full_name = payload.get('sub')
```

Comment:
There is a standardized claim called "name" which as I understand it is meant to contain the full name of the user. Could we use that one instead?